File: cpp/kmp.cpp
#include <vector>
#include <string>

using namespace std;

vector<size_t> get_pi(const string& pattern) {
    size_t begin = 1, matched = 0, pattern_len = pattern.length();
    vector<size_t> pi(pattern_len, 0);
    // pi[i]: pattern[..i]에서 Prefix == Suffix의 최대 길이
    while (begin + matched < pattern_len) {
        if (pattern[begin + matched] == pattern[matched]) {
            matched += 1;
            pi[begin + matched - 1] = matched;
        } else {
            if (matched == 0) begin += 1;
            else {
                begin += matched - pi[matched - 1];
                matched = pi[matched - 1];
            }
        }
    }
    return pi;
}

vector<size_t> kmp(const string& txt, const string& pattern) {
    size_t txt_len = txt.length(), pattern_len = pattern.length();
    vector<size_t> pi = get_pi(pattern);
    vector<size_t> found;
    size_t begin = 0, matched = 0;

    while (begin <= txt_len - pattern_len) {
        if (matched < pattern_len && txt[begin + matched] == pattern[matched]) {
            matched += 1;
            if (matched == pattern_len) found.push_back(begin);
        } else {
            if (matched == 0) begin += 1;
            else {
                begin += matched - pi[matched - 1];
                matched = pi[matched - 1];
            }
        }
    }

    return found;
}

File: rust/src/kmp.rs
use std::vec;

fn get_pi(s: &[u8]) -> Vec<usize> {
    let s_len = s.len();
    let mut pi = vec![0; s_len];
    let mut matched = 0;

    // https://yiyj1030.tistory.com/495
    for begin in 1..s_len {
        while matched > 0 && s[begin] != s[matched] {
            matched = pi[matched - 1];
        }
        if s[begin] == s[matched] {
            matched += 1;
            pi[begin] = matched;
        }
    }
    pi
}

fn kmp(s: &[u8], p: &[u8]) -> Vec<usize> {
    let pi = get_pi(p);
    let p_len = p.len();
    let s_len = s.len();
    let mut matched = 0;
    let mut find = vec![];

    for begin in 0..s_len {
        while matched > 0 && (matched < p_len && s[begin] != p[matched]) {
            matched = pi[matched - 1];
        }
        if matched < p_len && s[begin] == p[matched] {
            matched += 1;
            if matched == p_len {
                find.push(begin - p_len + 1);
                matched = pi[matched - 1];
            } 
        }
    }
    find
}